performant-software/Neatline
# zoom level and extent limits (#310 rebased) plus image-only exhibits

### What does this PR do?
This PR rebases #310 by @krdyke, "Allow user to set min/max zoom levels and restrict extent for base layer." It adds minor adjustments to cover edge case problems when loading certain layer types and and to account for zoom levels relative to the minimum level, as well as schema migrations.

It also adds the option to select no spatial layer for an exhibit, allowing users to have an image or WMS layer be the first to display.

Note: this bumps the Neatline version to 2.6.0, the anticipated next release version, in the interest of convenience for test installations.

### What issues does this address?
Closes #310, closes #295, closes #430 

### How to test
- On [staging](https://neatline-staging.herokuapp.com), go to Admin->Neatline and open a new or existing exhibit to the editor interface
- Click the Styles tab in the editor sidebar. Try adding values both manually and with the "Use Current" buttons to all fields in various combinations, saving and reloading at various points. Test zooming and panning on the map to confirm constraints and defaults have both taken effect correctly after a save+reload.
- Return to the Admin->Neatline page and click Exhibit Settings to switch the base layer type (test at least one Google layer, one Stamen layer and OpenStreetMaps), confirming that zoom/pan style settings work consistently across the different layer types.
- Also in Exhibit Settings, add a WMS layer (try @krdyke's test case of "http://nsidc.org/cgi-bin/atlas_north" for WMS Address and "land" for WMS Layers); test that this behaves correctly with the map in various constraint configurations.
### How to test, part 2:
- On [staging](https://neatline-staging.herokuapp.com), create a new exhibit. In the "Default Spatial Layer" dropdown, select "None (Image or WMS as Default)". Add an image url to the "Image Layer" field, e.g. https://upload.wikimedia.org/wikipedia/commons/e/e6/Roman_provinces_in_a_1867_school_atlas.jpg
- After saving the exhibit, load its public and editor views to check that the image displays correctly. The image should be centered in the viewer.
- Try setting default focus, default zoom, min/max zoom, and max extent values on the exhibit's style editor as described for spatial layers above. Reload to confirm that various combinations of these settings work as expected.
